app/scripts/controllers/selected-network-controller.ts

```diff
@@ -0,0 +1,154 @@
+  | GetSelectedNetworkState
+  | GetNetworkClientIdForDomain
+  | SetNetworkClientIdForDomain;
+
+export type SelectedNetworkControllerEvents = GetSelectedNetworkStateChange;
+
+export type SelectedNetworkControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  SelectedNetworkControllerActions,
+  NetworkControllerStateChangeEvent | SelectedNetworkControllerEvents,
+  never,
+  NetworkControllerStateChangeEvent['type']
+>;
+
+export type SelectedNetworkControllerOptions = {
+  messenger: SelectedNetworkControllerMessenger;
+};
+
+/**
+ * Controller for getting and setting the network for a particular domain.
+ */
+export default class SelectedNetworkController extends BaseControllerV2<
+  typeof controllerName,
+  SelectedNetworkControllerState,
+  SelectedNetworkControllerMessenger
+> {
+  /**
+   * Construct a SelectedNetworkController controller.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The restricted controller messenger for the EncryptionPublicKey controller.
+   */
+  constructor({ messenger }: SelectedNetworkControllerOptions) {
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: getDefaultState(),
+    });
+    this.registerMessageHandlers();
+  }
+
+  private registerMessageHandlers(): void {
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:getNetworkClientIdForDomain` as const,
+      this.getNetworkClientIdForDomain.bind(this),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:setNetworkClientIdForDomain` as const,
+      this.setNetworkClientIdForDomain.bind(this),
+    );
+
+    // subscribe to networkController statechange:: selectedNetworkClientId changed
+    // update the value for the domain 'metamask'
+    this.messagingSystem.subscribe(
+      'NetworkController:stateChange',
+      (state: NetworkState, patch: Patch[]) => {
+        const isChangingNetwork = patch.find(
+          (p) => p.path[0] === 'selectedNetworkClientId',
+        );
+        if (!isChangingNetwork) {
+          return;
+        }
+
+        // set it for the 'global' network to preserve functionality for the
+        // selectedNetworkController.perDomainNetwork feature flag being off
+        this.setNetworkClientIdForMetamask(state.selectedNetworkClientId);
+      },
+    );
+  }
+
+  /**
+   * Reset the controller state to the initial state.
+   */
+  resetState() {
+    this.update(getDefaultState);
+  }
+
+  setNetworkClientIdForMetamask(networkClientId: NetworkClientId) {
+    return this.setNetworkClientIdForDomain(METAMASK_DOMAIN, networkClientId);
+  }
+
+  setNetworkClientIdForDomain(
+    domain: Domain,
+    networkClientId: NetworkClientId,
+  ) {
+    this.update((state) => {
+      state.domains[domain] = networkClientId;
+    });
+  }
+
+  getNetworkClientIdForDomain(domain: Domain) {
+    if (this.state.perDomainNetwork === true) {
```

Comment:
would loosening to truthy be okay? `if (this.state.perDomainNetwork) {`

Reply:
yeah, it would, I just have a habbit of being more specific... mostly stemming from the days before typescript. Since this is a boolean, I think that would be fine